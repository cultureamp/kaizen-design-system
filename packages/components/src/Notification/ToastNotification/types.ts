import { HTMLAttributes } from "react"
import { NotificationType } from "../types"

type Modify<T, R> = Omit<T, keyof R> & R

export type ToastNotification = {
  id: string
  type: NotificationType
  title: string
  message: React.ReactNode
  onHide?: () => void
  /**
   * If this is true you will need to manage a way to remove the notification
   * @default false
   */
  persistent?: boolean
} & {
  [key: `data-${string}`]: unknown
}

export type ToastNotificationWithOptionals = Modify<
  ToastNotification,
  {
    id?: string
  }
>

export type AddToastNotification = (
  notification: ToastNotificationWithOptionals
) => void

export type RemoveToastNotification = (notificationId: string) => void

export type ClearToastNotifications = () => void
