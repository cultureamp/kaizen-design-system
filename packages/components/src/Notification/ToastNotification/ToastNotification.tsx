import React, { useId } from "react"
import { addToastNotification } from "./subcomponents/ToastNotificationManager"
import { ToastNotificationWithOptionals } from "./types"

export type ToastNotificationProps = Omit<
  ToastNotificationWithOptionals,
  "message" | "persistent"
> & {
  /**
   * Removes the dismiss trigger. functions the same as `persistent` in `addToastNotification`. If this is true you will need to manage the removal of notifications manually.
   * @default false
   */
  hideCloseIcon?: boolean
  children: React.ReactNode
}

export const ToastNotification = ({
  id: propsId,
  hideCloseIcon = false,
  type,
  title,
  onHide,
  children,
  ...restProps
}: ToastNotificationProps): null => {
  const reactId = useId()
  const id = propsId || reactId
  const persistent = hideCloseIcon

  addToastNotification({
    id,
    type,
    title,
    message: children,
    persistent,
    onHide,
    ...restProps,
  })

  return null
}

ToastNotification.displayName = "ToastNotification"
