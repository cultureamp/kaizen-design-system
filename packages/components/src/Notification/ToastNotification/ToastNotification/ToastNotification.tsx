import React, { useEffect, useId } from 'react'
import {
  GenericNotificationType,
  GenericNotificationVariant,
} from '~components/Notification/subcomponents/GenericNotification'
import { useToastNotificationContext } from '../context/ToastNotificationContext'
import { ToastNotificationObjBase } from '../types'

export type ToastNotificationProps = Omit<
  ToastNotificationObjBase,
  'id' | 'message' | 'persistent'
> & {
  children: React.ReactNode
  id?: string
  /**
   * Removes the dismiss trigger. Functions the same as `persistent` in `addToastNotification`.
   * If this is true you will need to manage the removal of notifications manually.
   * @default false
   */
  hideCloseIcon?: boolean
} & (GenericNotificationType | GenericNotificationVariant)

export const ToastNotification = ({
  id: propsId,
  hideCloseIcon = false,
  title,
  onHide,
  children,
  ...restProps
}: ToastNotificationProps): null => {
  const reactId = useId()
  const id = propsId ?? reactId
  const { addToastNotification } = useToastNotificationContext()
  const persistent = hideCloseIcon

  useEffect(() => {
    addToastNotification({
      id,
      title,
      message: children,
      persistent,
      onHide,
      ...restProps,
    })
  // Only run the callback once (proper fix would likely be a refactor)
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return null
}

ToastNotification.displayName = 'ToastNotification'
