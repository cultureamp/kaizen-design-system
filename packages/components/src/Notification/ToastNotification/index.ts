export * from "./ToastNotification"
export * from "./ToastNotificationsPortal"
export * from "./context/ToastNotificationContext"
export * from "./types"
