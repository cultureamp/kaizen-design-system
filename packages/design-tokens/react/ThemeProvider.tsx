import React from "react"
import { ThemeManager, defaultTheme, Theme } from "../"

export const ThemeContext = React.createContext<Theme>(defaultTheme)

/**
 * Wrap your application in this provider using a ThemeManager, to synchronise it with a react context.
 * This allows child react elements to more easily use theme variables, using the {@link useTheme} hook.
 */
export const ThemeProvider = ({
  themeManager,
  ...props
}: {
  themeManager: ThemeManager
  children: React.ReactNode
}) => {
  const [theme, setTheme] = React.useState<Theme>(
    themeManager.getCurrentTheme()
  )
  React.useEffect(() => {
    let cancelled = false
    const listener = (newTheme: Theme) => {
      if (!cancelled) setTheme(newTheme)
    }
    themeManager.addThemeChangeListener(listener)
    return () => {
      cancelled = true
      themeManager.removeThemeChangeListener(listener)
    }
  }, [])

  return (
    <ThemeContext.Provider value={theme}>
      {props.children}
    </ThemeContext.Provider>
  )
}

export const useTheme = () => React.useContext(ThemeContext)
