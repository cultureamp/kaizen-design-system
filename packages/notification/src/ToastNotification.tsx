import React from "react"
import { v4 } from "uuid"
import { addToastNotification } from "./ToastNotificationManager"
import { ToastNotificationWithOptionals } from "./types"

export type ToastNotificationProps = Omit<
  ToastNotificationWithOptionals,
  "message"
> & {
  hideCloseIcon?: boolean
  children: React.ReactNode
}

/**
 * {@link https://cultureamp.design/components/toast-notification/ Guidance} |
 * {@link https://cultureamp.design/storybook/?path=/docs/components-notification-toast-notification--cautionary Storybook}
 */
export const ToastNotification = ({
  id,
  hideCloseIcon = false,
  autohide = false,
  autohideDelay = "long",
  ...otherProps
}: ToastNotificationProps): null => {
  const [localID] = React.useState(id || v4())
  const persistent = autohide && hideCloseIcon

  addToastNotification({
    id: localID,
    type: otherProps.type,
    title: otherProps.title,
    automationId: otherProps.automationId,
    autohide,
    autohideDelay,
    message: otherProps.children,
    persistent,
    onHide: otherProps.onHide,
  })

  return null
}

ToastNotification.displayName = "ToastNotification"
