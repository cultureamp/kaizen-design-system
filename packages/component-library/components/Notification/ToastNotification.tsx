import * as React from "react"
import uuid from "uuid/v4"
import createNotificationManager from "./ToastNotificationManager"
import { NotificationType } from "./components/GenericNotification"

const notificationManager = createNotificationManager()

type Props = {
  type: NotificationType
  title: string
  children: React.ReactNode
  autohide: boolean
  autohideDelay?: "short" | "long"
  hideCloseIcon: boolean
  onHide?: () => void
  automationId?: string
}

const ToastNotification = ({ hideCloseIcon, ...otherProps }: Props) => {
  const [id] = React.useState(uuid())
  const persistent = otherProps.autohide && hideCloseIcon
  notificationManager.add({
    id,
    type: otherProps.type,
    title: otherProps.title,
    automationId: otherProps.automationId,
    autohide: otherProps.autohide,
    children: otherProps.children,
    persistent,
  })

  return null
}

ToastNotification.defaultProps = {
  autohide: false,
  hideCloseIcon: false,
  autohideDelay: "short",
}

export default ToastNotification
