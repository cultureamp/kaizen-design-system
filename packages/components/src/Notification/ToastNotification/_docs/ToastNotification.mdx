import { Canvas, Controls, Meta } from "@storybook/blocks"
import { ResourceLinks, Installation, LinkTo } from "~storybook/components"
import * as ToastNotificationStories from "./ToastNotification.stories"

<Meta title="Components/Notifications/ToastNotification" />

# Toast Notification

<ResourceLinks
  sourceCode="https://github.com/cultureamp/kaizen-design-system/tree/main/packages/components/src/Notification/ToastNotification"
  figma="https://www.figma.com/file/eZKEE5kXbEMY3lx84oz8iN/%F0%9F%92%9C-UI-Kit%3A-Heart?type=design&node-id=1929%3A21829&mode=design&t=QHnBiBn4stf5UxJ3-1"
  designGuidelines="https://cultureamp.atlassian.net/wiki/spaces/DesignSystem/pages/3075638310/Toast+Notification"
  className="!mb-8"
/>


<Installation
  installCommand="pnpm add @kaizen/components"
  importStatement='import { useToastNotification } from "@kaizen/components"'
/>

## Overview

Toast notifications are used to notifying a user of an outcome for an action they have taken.
They appended to the root of the document body and will render in the top corner of the window.

**Note:** Toast notifications must be used within <LinkTo pageId="components-kaizen-provider-installation">KaizenProvider</LinkTo>.

<Canvas of={ToastNotificationStories.Playground} />
<Controls of={ToastNotificationStories.Playground} />

## API

Use the following helper methods available through the `useToastNotification` hook to create and manage Toast notifications.

### addToastNotification

Adds and appends a new notification to the list.

<Canvas of={ToastNotificationStories.CreateNotification} />

### updateToastNotification

Updates a single notification that matches the `id` passed to the function.

<Canvas of={ToastNotificationStories.UpdateNotification} />

### removeToastNotification

Removes a single notification that matches the `id` passed to the function.

<Canvas of={ToastNotificationStories.RemoveNotification} />

### clearToastNotifications

Clears all notifications currently held in state.

Note that since each story has an individual context, this example will only clear the ones it adds.

<Canvas of={ToastNotificationStories.ClearNotifications} />

## Accessibility and removal of notifications

Any notification that is temporarily rendered on screen should provide a user ample time to read its contents and
provide a way to stop its removal ([WCAG guidline here](https://www.w3.org/TR/UNDERSTANDING-WCAG20/time-limits.html)).

By default, all Toast notifications will remain on screen until dismissed.
This ensures the accessibility is met and does not require additional intervention.

If you required to be automatically removed, you must provide a means of stopping and re-enabling its
removal and ensures that a user can reasonably access this before it is removed.
