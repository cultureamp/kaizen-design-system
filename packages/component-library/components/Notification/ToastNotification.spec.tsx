import { cleanup, render } from "@testing-library/react"
import * as React from "react"
import ToastNotification from "./ToastNotification"

afterEach(cleanup)

test("The basic notification renders correctly", () => {
  const { container } = render(
    <ToastNotification type="affirmative" title="Success">
      It worked!
    </ToastNotification>
  )

  expect(container.firstChild).toMatchSnapshot()
})
