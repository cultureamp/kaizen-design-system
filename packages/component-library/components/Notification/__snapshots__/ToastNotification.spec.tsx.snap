// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The basic notification renders correctly 1`] = `
<div
  class="notification affirmative toast hidden"
  data-automation-class="generic-notification affirmative toast"
  style="margin-top: 0px;"
>
  <div
    class="icon"
  >
    <svg
      aria-hidden="true"
      aria-labelledby="80284680-3304-5869-9788-7428e6c6ee86"
      class="icon inheritSize"
      focusable="false"
      role="presentation"
      viewBox="0 0 0 0"
    >
      <use
        xlink:href="#my-svg"
      />
    </svg>
  </div>
  <div
    class="textContainer"
  >
    <h6
      class="title"
    >
      Success
    </h6>
    <p
      class="text"
    >
      It worked!
    </p>
  </div>
  <button
    class="cancel"
    type="button"
  >
    <span
      class="cancelInner"
    >
      <svg
        aria-hidden="false"
        aria-labelledby="faffea72-0fc0-5bc3-8038-936bd9f22ca8"
        class="icon"
        focusable="false"
        role="img"
        viewBox="0 0 0 0"
      >
        <title
          id="faffea72-0fc0-5bc3-8038-936bd9f22ca8"
        >
          close notification
        </title>
        <use
          xlink:href="#my-svg"
        />
      </svg>
    </span>
  </button>
</div>
`;
